source/lib.ts

```diff
@@ -0,0 +1,312 @@
+// /source/lib.ts
+// The option parser and rate limiting middleware
+
+import Express from 'express'
+
+import {
+	Options,
+	AugmentedRequest,
+	RateLimitRequestHandler,
+	LegacyStore,
+	Store,
+	IncrementResponse,
+} from './types.js'
+import MemoryStore from './memory-store.js'
+
+/**
+ * Type guard to check if a store is legacy store.
+ *
+ * @param store {LegacyStore | Store} - The store to check
+ *
+ * @return {boolean} - Whether the store is a legacy store
+ */
+const isLegacyStore = (store: LegacyStore | Store): store is LegacyStore =>
+	// Check that `incr` exists but `increment` does not - store authors might want
+	// to keep both around for backwards compatibility.
+	typeof (store as any).incr === 'function' &&
+	typeof (store as any).increment !== 'function'
+
+/**
+ * Converts a legacy store to the promisified version.
+ *
+ * @param store {LegacyStore | Store} - The legacy store or even a modern store
+ *
+ * @returns {Store} - The promisified version of the store
+ */
+const promisifyStore = (passedStore: LegacyStore | Store): Store => {
+	if (!isLegacyStore(passedStore)) {
+		// It's not an old store, return as is
+		return passedStore
+	}
+
+	// Why can't Typescript understand this?
+	const legacyStore = passedStore
+
+	class PromisifiedStore implements Store {
+		async increment(key: string): Promise<IncrementResponse> {
+			return new Promise((resolve, reject) => {
+				legacyStore.incr(
+					key,
+					(
+						error: Error | undefined,
+						totalHits: number,
+						resetTime: Date | undefined,
+					) => {
+						if (error) reject(error)
+						resolve({ totalHits, resetTime })
+					},
+				)
+			})
+		}
+
+		async decrement(key: string): Promise<void> {
+			return Promise.resolve(legacyStore.decrement(key))
```

Comment:
I *think* that marking it as `async` is enough, you don't have to make it explicitly return a promise. I could be wrong though. We should test this with a couple of external stores just to be safe.

Reply:
Yes, marking it `async` is enough, I usually mark it explicitly `async` so if we change the code in the future and remove the `return Promise...`, it still does return a `Promise`.